subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmPluginExtension.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.plugins;
+
+import org.gradle.api.Action;
+import org.gradle.api.Incubating;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.plugins.jvm.JvmVariantBuilder;
+import org.gradle.api.plugins.jvm.JvmEcosystemUtilities;
+import org.gradle.api.plugins.jvm.OutgoingElementsBuilder;
+import org.gradle.api.plugins.jvm.ResolvableConfigurationBuilder;
+
+/**
+ * This extension is shared by JVM plugins (Java, Groovy, Scala, ...)
+ * and can be used to configure JVM specific behavior.
+ *
+ * @since 6.7
+ */
+@Incubating
+public interface JvmPluginExtension {
+    /**
+     * Provides access to the several handy JVM related utilities.
+     */
+    JvmEcosystemUtilities getUtilities();
+
+    /**
+     * Creates an outgoing configuration and configures it with reasonable defaults.
+     * @param name the name of the outgoing configurtion
+     * @param configuration the configuration builder, used to describe what the configuration is used for
+     * @return an outgoing (consumable) configuration
+     */
+    Configuration createOutgoingElements(String name, Action<? super OutgoingElementsBuilder> configuration);
```

Comment:
❓ Did we consider using `register` instead of `create` here, returning a `NamedDomainObjectProvider<Configuration>`?

Accessors for configurations and source sets in the Kotlin DSL are generated as providers that you can configure lazily. Configurations or source sets aren't lazily created at the moment, but they could be at some point. Uniformity of the APIs working with containers could also be an argument.

Reply:
I didn't do this because it would be a big lie at this stage. You'd call `register` and what it does is effectively creating a configuration. Part of the problem is this whole `maybeCreate` issue which is extremely common with `Configuration`: a plugin may create a configuration and you don't know, but if it doesn't you need configuration. There's also the fact that despite using `configurations.register`, configurations always end up being eagerly created (at this stage I didn't investigate why).

Reply:
Right, configuration creation isn't lazy. What about returning a provider for API uniformity even if the underlying container eagerly creates instances? There are plans to deprecate eager container APIs, I guess at first for tasks but any container should be concerned eventually. Returning a provider now would leave the possibility to change that without disturbing consumers of these new APIs.

Right, part of the problem is the `maybeCreate` issue, and the absence of a `maybeRegister` too.

Reply:
> What about returning a provider for API uniformity even if the underlying container eagerly creates instances?

I don't think it's a good idea. `register` means lazy. If we start adding confusion in that some `register` calls may be eager, then we're not improving UX imo. Note that we have plans to remove use of `Configuration` altogether from the public facing API, so this shouldn't be a problem in the (long) run.